astropy/modeling/statistic.py

```diff
@@ -6,36 +6,163 @@
+        Model with parameters set by the current iteration of the optimizer.
+        when evaluated on "x", must return array of shape "measured_vals"
+    weights : |ndarray| or None
+        Array of weights to apply to each residual.
+    *x : |ndarray|
+        Independent variables on which to evaluate the model.
+
+    Returns
+    -------
+    res : float
+        The sum of least squares.
+
+    Raises
+    ------
+    ValueError
+        If shape mismatch between model and measured
+    Exception
+        If cannot evaluate model, like because incorrect number of inputs
+        or a shape broadcasting failure.
+
+    See Also
+    --------
+    :func:`~astropy.modeling.statistic.leastsquare_1d`
+    :func:`~astropy.modeling.statistic.leastsquare_2d`
+    :func:`~astropy.modeling.statistic.leastsquare_3d`
+
+    Notes
+    -----
+    Models in :mod:`~astropy.modeling` have broadcasting rules that try to
+    match inputs with outputs with Model shapes. Numpy arrays have flexible
+    broadcasting rules, so mismatched shapes can often be made compatible. To
+    ensure data matches the model we must perform shape comparison and leverage
+    the Numpy arithmetic functions. This can obfuscate arithmetic computation
+    overrides, like with Quantities. Implement a custom statistic for more
+    direct control.
 
-def leastsquare(measured_vals, updated_model, weights, x, y=None):
     """
+
+    model_vals = updated_model(*x)
+
+    if np.shape(model_vals) != np.shape(measured_vals):
+        raise ValueError(
+            (
+                f"Shape mismatch between model ({np.shape(model_vals)})"
+                f" and measured ({np.shape(measured_vals)})"
+            )
+        )
+
+    if weights is None:
+        weights = 1.0
+
+    return np.sum(np.square(weights * np.subtract(model_vals, measured_vals)))
+
+
+# -------------------------------------------------------------------
+
+_leastsquare_XD_doc: str = """
     Least square statistic with optional weights.
+    Safer than the general :func:`~astropy.modeling.statistic.leastsquare`
+    for {XD} models by avoiding numpy methods that support broadcasting.
+
+    .. |ndarray| replace:: `~numpy.ndarray`
 
     Parameters
     ----------
-    measured_vals : `~numpy.ndarray`
+    measured_vals : |ndarray|
         Measured data values.
     updated_model : `~astropy.modeling.Model`
         Model with parameters set by the current iteration of the optimizer.
-    weights : `~numpy.ndarray`
+    weights : |ndarray| or None
         Array of weights to apply to each residual.
-    x : `~numpy.ndarray`
-        Independent variable "x" to evaluate the model on.
-    y : `~numpy.ndarray`, optional
-        Independent variable "y" to evaluate the model on, for 2D models.
+    {parameters}
 
     Returns
     -------
     res : float
         The sum of least squares.
+
+    See Also
+    --------
+    :func:`~astropy.modeling.statistic.leastsquare`
+
+"""
```

Comment:
I don't know. I find the complexity of the docstring manipulations here don't justify the reduction of duplications. They are otherwise pretty simple even if a few lines are repeated in each (They could even be composed of docstring fragments and not rely on decorators). I'm curious if other reviewers agree.

Reply:
✅. I've just put in the whole docstring.